tower-http/src/trace.rs

```diff
@@ -0,0 +1,348 @@
+
+    fn layer(&self, inner: S) -> Self::Service {
+        Trace {
+            inner,
+            latency_unit: self.latency_unit,
+            classifier: self.classifier.clone(),
+            event_emitter: self.event_emitter.clone(),
+        }
+    }
+}
+
+/// Middleware that add high level [tracing] to a [`Service`].
+///
+/// [tracing]: https://crates.io/crates/tracing
+/// [`Service`]: tower_service::Service
+#[derive(Clone, Debug)]
+pub struct Trace<S, Classifier, EventEmitter> {
+    inner: S,
+    latency_unit: LatencyUnit,
+    classifier: Classifier,
+    event_emitter: EventEmitter,
+}
+
+impl<ReqBody, ResBody, S, Classifier, EventEmitter> Service<Request<ReqBody>>
+    for Trace<S, Classifier, EventEmitter>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    Classifier: ClassifyResponse<ResBody, S::Error> + Clone,
+    EventEmitter: EmitTracingEvents<ReqBody, OkClass = Classifier::OkClass, ErrClass = Classifier::ErrClass>
+        + Clone,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future, Classifier, EventEmitter, ReqBody>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let start = Instant::now();
+
+        let span = self.event_emitter.make_span(&req);
+
+        let future = {
+            let _guard = span.enter();
+            self.event_emitter.on_request_received(&req);
+            self.inner.call(req)
+        };
+
+        ResponseFuture {
+            future,
+            span,
+            start,
+            latency_unit: self.latency_unit,
+            classifier: self.classifier.clone(),
+            event_emitter: self.event_emitter.clone(),
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// The [`Future`] produced by [`Trace`] services.
+///
+/// [`Future`]: std::future::Future
+#[pin_project]
+#[derive(Debug)]
+pub struct ResponseFuture<F, Classifier, EventEmitter, ReqBody> {
+    #[pin]
+    future: F,
+    span: Span,
+    start: Instant,
+    latency_unit: LatencyUnit,
+    classifier: Classifier,
+    event_emitter: EventEmitter,
+    _marker: PhantomData<fn() -> ReqBody>,
+}
+
+impl<F, ReqBody, ResBody, E, Classifier, EventEmitter> Future
+    for ResponseFuture<F, Classifier, EventEmitter, ReqBody>
+where
+    F: Future<Output = Result<Response<ResBody>, E>>,
+    Classifier: ClassifyResponse<ResBody, E>,
+    EventEmitter:
+        EmitTracingEvents<ReqBody, OkClass = Classifier::OkClass, ErrClass = Classifier::ErrClass>,
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        let _guard = this.span.enter();
+
+        let result = ready!(this.future.poll(cx));
+        let latency = this.start.elapsed();
```

Comment:
it seems like this always records latency in terms of time until the beginning of the response. i can also see use-cases that might want to record latency in terms of time to first byte received, or time to response body EOS. we might want to think about allowing the user to customize this — but, it's probably out of scope for this initial PR; just something to think about later.

Reply:
Uhh yeah thats a good point. Didn't consider that.